File: libs/core/src/core/queue/contracts.py
"""Versioned pydantic job payloads for every Redis Stream (PRD §4.2).

Queue payloads are a wire contract between services that deploy
independently — hence pydantic validation at the edges and an explicit
``schema_version`` on every message, so an old consumer can reject (or
tolerate) a newer producer instead of misreading its fields.
"""

from __future__ import annotations

import uuid

from pydantic import BaseModel, Field

SCHEMA_VERSION = 1


class JobPayload(BaseModel):
    """Base for every queue message. ``schema_version`` bumps on any
    breaking change to a payload's fields."""

    schema_version: int = SCHEMA_VERSION


class SplitJob(JobPayload):
    """Stream ``doc.split`` — produced by the API on commit."""

    doc_id: uuid.UUID
    source_uri: str


class ParseJob(JobPayload):
    """Stream ``doc.parse`` — produced by the splitter, one per shard."""

    doc_id: uuid.UUID
    idx: int = Field(ge=0)
    page_start: int = Field(ge=1)
    page_end: int = Field(ge=1)
    source_uri: str

    def validate_pages(self) -> None:
        if self.page_end < self.page_start:
            raise ValueError(f"page_end {self.page_end} < page_start {self.page_start}")


class EmbedJob(JobPayload):
    """Stream ``doc.embed`` — produced when a document's last shard settles."""

    doc_id: uuid.UUID

File: internal/queue/contracts.go
// Package queue: versioned job payloads for every Redis Stream (PRD §4.2).
//
// Queue payloads are a wire contract between services that deploy
// independently — hence an explicit SchemaVersion on every message, so an
// old consumer can reject (or tolerate) a newer producer instead of
// misreading its fields. JSON tags are identical to 1.0's contracts.py.
package queue

import "encoding/json"

// SchemaVersion bumps on any breaking change to a payload's fields.
const SchemaVersion = 1

// SplitJob is the doc.split payload — produced by the API on commit.
type SplitJob struct {
	SchemaVersion int    `json:"schema_version"`
	DocID         string `json:"doc_id"`
	SourceURI     string `json:"source_uri"`
}

// ParseJob is the doc.parse payload — produced by the splitter, one per shard.
type ParseJob struct {
	SchemaVersion int    `json:"schema_version"`
	DocID         string `json:"doc_id"`
	Idx           int    `json:"idx"`
	PageStart     int    `json:"page_start"`
	PageEnd       int    `json:"page_end"`
	SourceURI     string `json:"source_uri"`
}

// ValidatePages enforces page_end >= page_start (1.0 contracts.py).
func (j *ParseJob) ValidatePages() error {
	if j.PageEnd < j.PageStart {
		return &PayloadError{Field: "page_end", Msg: "page_end < page_start"}
	}
	return nil
}

// EmbedJob is the doc.embed payload — produced when a document's last
// shard settles.
type EmbedJob struct {
	SchemaVersion int    `json:"schema_version"`
	DocID         string `json:"doc_id"`
}

// PayloadError marks a structurally invalid job payload.
type PayloadError struct {
	Field string
	Msg   string
}

func (e *PayloadError) Error() string { return e.Field + ": " + e.Msg }

// MarshalJob is the single serializer every producer uses.
func MarshalJob(v any) (string, error) {
	b, err := json.Marshal(v)
	if err != nil {
		return "", err
	}
	return string(b), nil
}
